app/javascript/context/Reducer.js

```diff
@@ -1,97 +1,44 @@
-/**
- * Action Definitions below. Actions should be named ACTION_<NAME> where
- * <NAME> is a noun describing an event that happened (past tense).
- */
-export const ACTION_TYPE_SELECTED_LISTING_CHANGED = 'SET_CURRENT_LISTING'
-export const ACTION_TYPE_SELECTED_APPLICATION_CHANGED = 'SET_CURRENT_APPLICATION'
-export const ACTION_TYPE_LEFT_LISTING_SCOPE = 'CLEAR_CURRENT_LISTING'
-export const ACTION_TYPE_LEFT_APPLICATION_SCOPE = 'CLEAR_CURRENT_APPLICATION'
-export const ACTION_TYPE_APPLICATION_LOADED = 'SUPP_APPLICATION_LOADED'
-
-export const ACTION_TYPE_APPLICATION_TABLE_FILTERS_APPLIED = 'APPLICATION_TABLE_FILTERS_APPLIED'
-export const ACTION_TYPE_APPLICATION_TABLE_PAGE_CHANGED = 'APPLICATION_TABLE_PAGE_CHANGED'
-
-export const getEmptyListing = () => ({
-  id: null,
-  name: null,
-  buildingAddress: null
-})
+import ApplicationDetailsSubreducer from 'context/subreducers/ApplicationDetailsSubreducer'
+import ApplicationsListSubreducer from 'context/subreducers/ApplicationsListSubreducer'
+import BreadcrumbSubreducer from 'context/subreducers/BreadcrumbSubreducer'
+
+// If a new sub-reducer is added, you should just need to add it to this object
+// and it will be integrated into the root reducer automatically.
+const DATA_KEYS_TO_SUB_REDUCERS = {
+  breadcrumbData: BreadcrumbSubreducer,
+  applicationsListData: ApplicationsListSubreducer,
+  applicationDetailsData: ApplicationDetailsSubreducer
+}
 
-export const getEmptyApplication = () => ({
-  id: null,
-  number: null,
-  applicantFullName: null
-})
+const mapEachSubReducer = (mapFn) => {
+  return Object.fromEntries(
+    Object.entries(DATA_KEYS_TO_SUB_REDUCERS).map(([key, subReducer]) => [
+      key,
+      mapFn(subReducer, key)
+    ])
+  )
+}
 
-export const getEmptyApplicationsListData = () => ({
-  appliedFilters: {},
-  page: 0
-})
+export const getInitialState = () => mapEachSubReducer((r) => r.getInitialState())
+export const anyReducerHandlesAction = (actionType) =>
+  Object.values(DATA_KEYS_TO_SUB_REDUCERS).some((r) => r.handlesActionType(actionType))
 
-const overrideBreadcrumbData = (state, { listing, application }) => ({
-  ...state,
-  breadcrumbData: {
-    ...state.breadcrumbData,
-    ...(listing && {
-      listing: {
-        id: listing.id,
-        name: listing.name,
-        buildingAddress: listing.buildingAddress
-      }
-    }),
-    ...(application && {
-      application: {
-        id: application.id,
-        number: application.number,
-        applicantFullName: application.applicantFullName
-      }
-    })
-  }
-})
-
-const overrideApplicationPageData = (state, { appliedFilters, page }) => ({
-  ...state,
-  applicationsListData: {
-    ...state.applicationsListData,
-    ...(appliedFilters && { appliedFilters }),
-    ...(!(page === undefined || page === null) && { page })
+const Reducer = (state, action) => {
```

Comment:
Basically what this does is:
1. When an action comes in, see what subreducers can handle the action type (can be multiple)
2. Call the applicable subreducers and set their result state to the data key associated
3. return the entire state object, with each top-level key (breadcrumbData, applicationDetailsData, etc.) referring to a subreducer's state